### ISSUE TITLE
Optimizing model parameters and training outputs

### ISSUE THREAD

**Issue by chitayata** [janclemenslab/das#42]
Dear Jan and team, 

I am using das through Python and would like to fit the network with different configurations to compare performance and find the best model. 
I see there is a shell script for the command-line on the terminal [(janclemenslab: TRAIN)](https://janclemenslab.org/das/tutorials/train.html) but is there a way to do this with the train function in the das.train module in python?

Additionally, how can I get python to print the training outputs? 
If I run my training in the GUI, I get a printout like below that includes the f1-scores.

### INFO:das.train:{'noise': {'precision': 0.9175990751370318, 'recall': 0.979656218576971, 'f1-score': 0.9476127362477954, 'support': 231717}, 'pulse': {'precision': 0.05627009646302251, 'recall': 0.015928398058252427, 'f1-score': 0.024828564672499408, 'support': 6592}, 'sine': {'precision': 0.8592689767483941, 'recall': 0.5747178602765423, 'f1-score': 0.6887611726526097, 'support': 33051}, 'accuracy': 0.9069243808962264, 'macro avg': {'precision': 0.6110460494494828, 'recall': 0.5234341589705885, 'f1-score': 0.5537341578576348, 'support': 271360}, 'weighted avg': {'precision': 0.8895708148582068, 'recall': 0.9069243808962264, 'f1-score': 0.8936685429716721, 'support': 271360}}

But I don't get this python. I am using the code below with the same files (including test set) and parameters as the GUI. Am I missing an argument to print the results?

`model, params = das.train.train(model_name='tcn',  
                                data_dir='Quickstart3.npy', 
                                save_dir='Quickstart3.res',
                                nb_hist= 256,
                                kernel_size=16,
                                nb_filters=16,
                                batch_size=16,
                                ignore_boundaries=True,
                                verbose=1,
                                nb_epoch=4,
                                log_messages= True)`

Any advice is much appreciated!

**Comment by chitayata** [comment]
Just a quick update.

I wasn't able to get the results to print automatically, however, I was able to use the following code to load and show the '_results.h5' file created during training. 

```
import flammkuchen
import logging
logging.basicConfig(level=logging.INFO)
res = flammkuchen.load(f'pathway_results.h5')    #Load the results.h5 file
res   #print results 
```

Still no luck with parameter optimization within python though...

**Comment by postpop** [comment]
Hi, glad you figured out a way to access the test results.

Regarding the optimization: The search space is not that big - only 3x3x2=18 combinations. You could brute-force this and try all combinations using `das.train.train` and a for loop.

We also have experimental support for automatic parameter tuning via [keras tuner](https://keras.io/keras_tuner/) in `das.train_tune`. The interface is similar to `das.train`. 

Python: `das.train_tune.train(data_dir='tutorial_data.npy', save_dir='res', kernel_size=3, tune_config='tune.yml')`. There is also a CLI that you can access via `das tune`.

Crucially, it accepts a yaml file with the parameter names and values you want to optimize. In your case, the `tune.yml` file would look like this:
```yaml
nb_conv: [2, 3, 4]
nb_filters: [32, 64, 96]
learning_rate: [0.0001, 0.00001]
```
The tuner will then run a bunch of fits and find an optimal parameter combination in this search space - see [keras tuner](https://keras.io/keras_tuner/) for how this works.

But again, in your case I think it would be easier to just run all 18 fits a couple of times to optimize these parameters.

Good luck and let me know how this goes. In particular, if you use the tuner and run into issues, because we have so far only used this internally. Also happy to have a look at your data with you to give some advice on the model parameters.

**Comment by chitayata** [comment]
Hi Jan, 

Great, I will first try a for loop, but may also try the automatic tuner to see how it works and if I end up with different results.  I will let you know how it goes. I am occasionally running into memory iussues however, so hopefully it won't be a problem to run. 

It would be great to your input on model parameters though! I will send an email with my data to the one listed your site. 

Many thanks for your suggestions.

**Comment by postpop** [comment]
Thanks for giving this a try. 

Looks like it just fails to train - loss is nan. Does the same model work when using the regular training?

Something like this?
```python
das.train_tune.train(data_dir='normalized_FR.npy', save_dir='normalized_FR.res', kernel_size=16, nb_epoch=4, nb_filters=32)
```

**Comment by postpop** [comment]
I'll test this on my side as well just to make sure.

**Comment by chitayata** [comment]
When running the code provided I had the same issue as before - loss is nan.
`das.train_tune.train(data_dir='normalized_FR.npy', save_dir='normalized_FR.res', kernel_size=16, nb_epoch=4, nb_filters=32)`

While it didn't work, I could see it was attempting to run multiple trails though. So just out of an interest to understand the code, why did it do this? I thought because we specified an exact value for each argument that it would just run 1 trail. 
Or is it conducting a random search?

When you said 'regular training' though did you mean `das.train.train()`?
`das.train.train(data_dir='normalized_FR.npy', save_dir='normalized_FR.res', kernel_size=16, nb_epoch=4, nb_filters=32)`

Running this code worked as it should.

**Comment by postpop** [comment]
Oh, yes, I meant `das.train.train`, sorry about the confusion.

The tuning works by running different trials: it fits models with different parameters. Based on the results of the current trial,  the most promising set of parameters for the next trial is selected. So it's doing something a bit smarter than a random search. 

And I also get nan loss now - probably some change in keras tuner that breaks things in DAS. I'll fix this and let you know once it works.

**Comment by chitayata** [comment]
That's a really nice feature. So the argument inputs we specify just give it a starting point and then it optimizes from there.

Ah ok, thanks for trying it. Sounds good.

**Comment by postpop** [comment]
I've updated `das.tune.train` to now work with the new keras tuner API. 
You can give this a try by updating to the latest version 0.28.0 via pip (it's not on conda yet):
`pip install das --upgrade --no-deps`

**Comment by chitayata** [comment]
Fabulous thank you so much! keep you posted

**Comment by chitayata** [comment]
Got it updated and was able to run it it with a .yml file, providing the specific values I wanted to run test and compare. 
Seems to be running well and will be helpful for optimization. 
I find this easier than creating a for loop and I think I will stick with it. 
Many thanks!

**Comment by postpop** [comment]
Great - closing this.
